metagraph/src/graph/representation/canonical_dbg.cpp

```diff
@@ -172,36 +185,57 @@ void CanonicalDBG::append_next_rc_nodes(node_index node,
 
     // for each n, check for nAGCCA. If found, define and store the index for
     // TGGCTrc(n) as index(nAGCCA) + offset_
-    if (const auto *dbg_succ = dynamic_cast<const DBGSuccinct*>(graph_.get())) {
+    if (const auto *dbg_succ = get_dbg_succ(*graph_)) {
+        const auto *cached = dynamic_cast<const DBGSuccinctCachedView*>(graph_.get());
         const auto &boss = dbg_succ->get_boss();
-        dbg_succ->call_nodes_with_suffix_matching_longest_prefix(
-            std::string_view(&rev_seq[1], get_k() - 1),
-            [&](node_index next, uint64_t /* match length */) {
-                auto edge = dbg_succ->kmer_to_boss_index(next);
-                boss::BOSS::TAlphabet c = boss.get_minus_k_value(edge, get_k() - 2).first;
-                if (c == boss::BOSS::kSentinelCode)
-                    return;
 
-                c = kmer::KmerExtractorBOSS::complement(c);
+        auto next_callback = [&](node_index next, uint64_t /* match length */) {
+            auto edge = dbg_succ->kmer_to_boss_index(next);
+            boss::BOSS::TAlphabet c = cached
+                ? cached->get_first_value(edge)
+                : boss.get_minus_k_value(edge, get_k() - 2).first;
+            if (c == boss::BOSS::kSentinelCode)
+                return;
 
-                if (children[c] == npos) {
-                    children[c] = next + offset_;
-                    return;
-                }
+            rev_seq[0] = boss.decode(c);
+            c = kmer::KmerExtractorBOSS::complement(c);
 
-                if (k_odd_) {
-                    logger->error(
-                        "Primary graph contains both forward and reverse complement: {} {} -> {} {}\t{} {}",
-                        node, graph_->get_node_sequence(node),
-                        children[c], graph_->get_node_sequence(children[c]),
-                        next, graph_->get_node_sequence(next));
-                    exit(1);
-                }
+            if (children[c] == npos) {
+                children[c] = next + offset_;
+                if (cached)
+                    cached->put_decoded_node(next, rev_seq);
```

Comment:
Why does Canonical wrapper care about the cache? Doesn't it have its own cache anyway?

Reply:
Doesn't it make sense to always keep a cached view on the graph in CanonicalDBG, so there will be no question if this thing is cached or not?

Reply:
Caching is only supported for DBGSuccinct, whereas CanonicalDBG is more general. I guess one way around this is to use CanonicalDBG<DBGSuccinctCachedView> instead of CanonicalDBG<DeBruijnGraph>, so these if-statements can be made constexpr

Reply:
huh....I just noticed now that the `<....>` were discarded by GitHub's markdown :P

Reply:
anyway, I cleaned this part up a little. Changing the internals of CanonicalDBG to be templated is something that I think we can save for another PR